pkg/template/controller/templateinstance_finalizer.go

```diff
@@ -0,0 +1,261 @@
+		readinessLimiter:  workqueue.NewItemFastSlowRateLimiter(5*time.Second, 20*time.Second, 200),
+		clock:             clock.RealClock{},
+	}
+
+	informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			t := obj.(*templateapi.TemplateInstance)
+			if t.DeletionTimestamp != nil {
+				c.enqueue(t)
+			}
+		},
+		UpdateFunc: func(_, obj interface{}) {
+			t := obj.(*templateapi.TemplateInstance)
+			if t.DeletionTimestamp != nil {
+				c.enqueue(t)
+			}
+		},
+	})
+
+	c.jsonConfig = rest.CopyConfig(config)
+	c.jsonConfig.ContentConfig = dynamic.ContentConfig()
+	var err error
+	c.client, err = dynamic.NewForConfig(c.jsonConfig)
+	if err != nil {
+		glog.Errorf("failure creating dynamic client: %v", err)
+		return nil
+	}
+
+	return c
+}
+
+// getTemplateInstance returns the TemplateInstance from the shared informer,
+// given its key (dequeued from c.queue).
+func (c *TemplateInstanceFinalizerController) getTemplateInstance(key string) (*templateapi.TemplateInstance, error) {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.lister.TemplateInstances(namespace).Get(name)
+}
+
+// sync is the actual controller worker function.
+func (c *TemplateInstanceFinalizerController) sync(key string) error {
+	templateInstance, err := c.getTemplateInstance(key)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	needsFinalizing := false
+	for _, v := range templateInstance.Finalizers {
+		if v == templateapi.TemplateInstanceFinalizer && templateInstance.DeletionTimestamp != nil {
+			needsFinalizing = true
+			break
+		}
+	}
+	if !needsFinalizing {
+		return nil
+	}
+
+	glog.V(4).Infof("TemplateInstanceFinalizer controller: syncing %s", key)
+
+	errs := []error{}
+	background := metav1.DeletePropagationBackground
+	deleteOpts := &metav1.DeleteOptions{PropagationPolicy: &background}
+	for _, o := range templateInstance.Status.Objects {
+		glog.V(5).Infof("attempting to delete object: %#v", o)
+
+		gv, err := schema.ParseGroupVersion(o.Ref.APIVersion)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error parsing group version %s for object %#v: %v", o.Ref.APIVersion, o, err))
+			continue
+		}
+		gk := schema.GroupKind{
+			Group: gv.Group,
+			Kind:  o.Ref.Kind,
+		}
+		mapping, err := c.dynamicRestMapper.RESTMapping(gk, gv.Version)
+		if err != nil || mapping == nil {
+			errs = append(errs, fmt.Errorf("error mapping object %#v: %v", o, err))
+			continue
+		}
+
+		namespaced := mapping.Scope.Name() == meta.RESTScopeNameNamespace
+		namespace := ""
+		if namespaced {
+			namespace = o.Ref.Namespace
+		}
+
+		gvr := schema.GroupVersionResource{
+			Group:    gv.Group,
+			Version:  gv.Version,
+			Resource: mapping.Resource,
+		}
+
+		var resourceClient dynamic.DynamicResourceInterface
+		if namespaced {
```

Comment:
this gets less ugly after the rebase.

Reply:
yeah i saw, i'll clean it up at that point i guess.